uwblueprint/social-diversity-for-children
# Mg/document upload

### Notion ticket link



[Ticket Name](https://www.notion.so/uwblueprintexecs/086c1b564dbf478faa649a8880e34b81?v=9d3f8002a0d1477ba395616dfc082af7&p=087294c860404202bd85708c7e5c6b8a)

https://user-images.githubusercontent.com/35342787/136640674-e92824af-82ed-48aa-a8c2-af77b045b074.mp4



### Implementation description

* Created 2 components
    * DragAndDrop.tsx
        * uses a FileUploadButton
        * **NO drag and drop functionality yet** will look into 3rd party package most likely https://www.npmjs.com/package/dropzone
    * FileUploadButton.tsx
        * Uses a button and an input (ref on input so when button clicked we add files to the input
        * Multiple files are supported if you uncomment the multiple keyword on the input (though formatting may get messed up)
* Created document-upload page
    * Checks sessions, redirects to home if no valid session
    * gets query parameter of `/document-upload?type=SOMETHING`
        *  valid types are "", "criminal-check", "income-proof", "curriculum-plans"
            * if no parameter, s3 bucket folder will be sent to "other"
            * if invalid parameter type, 404 returned in console (no error page yet) from backend
            * if query parameter is not `type` redirected back to `/document-upload`
            * if query parameter `type` is not set redirected to  `/document-upload`
     * gets presigned s3 url for posting from `/api/upload/url?path=TYPE}&file=PREFIXED_FILENAME`
        * filename currently prefixed with user info as `email-session_id-filename`
            * will transition from this prefix to another method after demo ([ticket](https://www.notion.so/uwblueprintexecs/086c1b564dbf478faa649a8880e34b81?v=9d3f8002a0d1477ba395616dfc082af7&p=c105d7c5ea61484991075d1f2aacd11a))
     * uploads file as `FormData` to presigned url
* Created backend api `/api/upload/url?path=TYPE}&file=PREFIXED_FILENAME`
    * checks valid session
    * checks valid upload types
    * creates presigned url with key (s3 path) from type and filename
    * max 5MB file currently
* Simple terraform configuration for s3 creation
    * creates folders in s3
    * upload types/folder names specified in `terraform.tfvars` --> decoupled from type checks in api backend
        *  would be good to centralize



### Steps to test

1. Try using an invalid query parameter (?xxxx=...) --> redirect back to `/document-upload`'
2. Try not assigning query parameter `?type` to anything --> redirected back to  `/document-upload`
2. Try uploading without logging in (should be redirected back to home)
3. Try uploading with invalid value not support query parameters (404 returned from backend)
2. Try uploading logging in
3. Check S3 bucket for file



### What should reviewers focus on?
* Dev: File uploading for supported types ("",  "criminal-check",  "income-proof",   "curriculum-plans")
* Dev: Logic of eligibility
* Dev: Code style
* Design: Look and feel of upload process

### TODO 
* route to 404 page
* drag and drop functionality
* better error handling
* store s3 filenames for each uploader in database
### Checklist
